convert2rhel/applock.py

```diff
@@ -0,0 +1,186 @@
+
+__metaclass__ = type
+
+import errno
+import io
+import logging
+import os
+import tempfile
+
+
+_DEFAULT_LOCK_DIR = "/var/run/lock"
+loggerinst = logging.getLogger(__name__)
+
+
+class ApplicationLockedError(Exception):
+    """Raised when this application is locked."""
+
+    def __init__(self, message):
+        super(ApplicationLockedError, self).__init__(message)
+        self.message = message
+
+
+class ApplicationLock:
+    """Holds a lock for a particular application.
+
+    To acquire and release the lock, we recommend using the context
+    manager, though try_to_lock() and unlock() methods are provided. You
+    may call unlock() without having called try_to_lock() first, so it
+    can be used in a cleanup routine without worries.
+
+    The implementation uses a standard Linux PID file. When a program
+    that uses ApplicationLock starts, it writes its process ID to a
+    file in /var/run/lock. If the file already exists, it reads the
+    PID and checks to see if the process is still running. If the
+    process is still running, it raises ApplicationLockedError. If the
+    process is not running, it removes the file and tries to lock it
+    again.
+
+    For safety, unexpected conditions, like garbage in the file or
+    bad permissions, are treated as if the application is locked,
+    because something is obviously wrong.
+    """
+
+    def __init__(self, name):
+        # Our application name
+        self._name = name
+        # Do we think we locked the pid file?
+        self._locked = False
+        # Maximum number of tries to lock
+        self._max_loop_count = 5
+        # Our process ID
+        self._pid = os.getpid()
+        # Path to the file that contains the process id
+        self._pidfile = os.path.join(_DEFAULT_LOCK_DIR, self._name + ".pid")
+
+    def __str__(self):
+        if self._locked:
+            status = "locked"
+        else:
+            status = "unlocked"
+        return "%s PID %d %s" % (self._pidfile, self._pid, status)
+
+    def _try_create(self):
+        """Try to create the lock file. If this succeeds, the lock file
+        exists and we created it. If an exception other than the one
+        we expect is raised, re-raises it.
+
+        :returns: True if we created the lock, False if we didn't.
+        """
+        #
+        # Create a temporary file that contains our PID and attempt
+        # to link it to the real pidfile location. The link will fail if
+        # the file already exists; this avoids a race condition when
+        # two processes attempt to create the file simultaneously. This
+        # also guarantees that the lock file contains valid data.
+        #
+        with tempfile.NamedTemporaryFile(mode="w", suffix=".pid", prefix=self._name, dir=_DEFAULT_LOCK_DIR) as f:
+            f.write(str(self._pid) + "\n")
+            f.flush()
+            try:
+                os.link(f.name, self._pidfile)
```

Comment:
@jochapma, not sure if its worth to comment here in the code, but I think that adding a comment like: `tempfile.NamedTemporaryFile will delete the temporary file created once the context manager exits, thus, we are linking the temp file created to self._pidfile`.

Do you think this would bring any value to other people reading the code?

Reply:
Added a comment.

Reply:
Great comment, btw!